### DISCUSSION TITLE
Easily unit/integration test your API routes with NTARH!

### DISCUSSION THREAD

**Discussion by Xunnamius** [vercel/next.js#60838] [Show and tell]
Hi there! I find I'm constantly starting API-heavy projects with Next.js, and as a TDD adherent, I thrive on the instant feedback unit testing provides. Yet I've found it difficult to test my Next.js API routes without rerunning time-consuming end-to-end tests or running through tedious manual testing with something like Postman. This lead me to develop [NTARH (next-test-api-route-handler)](https://github.com/Xunnamius/next-test-api-route-handler#readme) with the goal of improving the unit and integration testing DX for Next.js routes.

Here's an example of an app route tested with NTARH:

```typescript
/* File: test/unit.test.ts */

import { testApiHandler } from 'next-test-api-route-handler';
// Import the handler under test from the app directory
import * as edgeHandler from '../app/your-edge-endpoint/route';

it('does what I want', async function () {
  await testApiHandler<{ success: boolean }>({
    appHandler: edgeHandler,
    requestPatcher(request) {
      return new NextRequest(request, {
        geo: { city: 'Chicago', country: 'United States' },
        ip: '110.10.77.7'
      });
    },
    async test({ fetch }) {
      await expect((await fetch()).json()).resolves.toStrictEqual({
        success: true
      });
    }
  });
});
```

I've been using NTARH successfully for both personal projects and to test production deployments over the past few years, and we just landed App Router support with version 4. If you've got some API routes you're working on, [give it a spin](https://github.com/Xunnamius/next-test-api-route-handler#install). Feedback welcome :)

**Comment by kamtoeddy** [comment]
I am also a test freak and recently started a project where a custom backend would be a lot more work.
Loving NTARH so far.

Thanks a lot
